fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/FilterWithConjunctNestedToNestingContextRule.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * FilterWithConjunctNestedToNestingContextRule.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp.rules;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.query.expressions.AndComponent;
+import com.apple.foundationdb.record.query.expressions.NestedField;
+import com.apple.foundationdb.record.query.expressions.QueryComponent;
+import com.apple.foundationdb.record.query.plan.temp.SingleNestedContext;
+import com.apple.foundationdb.record.query.plan.temp.ExpressionRef;
+import com.apple.foundationdb.record.query.plan.temp.NestedContext;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRule;
+import com.apple.foundationdb.record.query.plan.temp.PlannerRuleCall;
+import com.apple.foundationdb.record.query.plan.temp.expressions.LogicalFilterExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.NestedContextExpression;
+import com.apple.foundationdb.record.query.plan.temp.expressions.RelationalPlannerExpression;
+import com.apple.foundationdb.record.query.plan.temp.matchers.AnyChildWithRestMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ReferenceMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.TypeMatcher;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * A rule that looks for a logical filter that contains a conjunction that includes a non-repeated nested predicate
+ * ({@link NestedField}) and transforms the inner expression to a version with respect to the {@link NestedContext}
+ * defined by that nested field. The other filters in the conjunct are left above the produced
+ * {@link NestedContextExpression} and need to be handled separately.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class FilterWithConjunctNestedToNestingContextRule extends PlannerRule<LogicalFilterExpression> {
```

Comment:
Suppose you had something like `Query.and(Query.field("parent").matches(something), Query.field("parent").matches(somethingElse), Query.field("parent").matches(aThirdThing))` where `parent` is non-repeated. I think, the way this rule is executed, this would transform the expression a few different times (essentially a different nested context for each of the parents), and it might do that a few different times.

Do you think we need to worry about the number of these things "exploding" in that case, or is it bounded in some way that I'm missing (or is my analysis wrong)? Maybe it's fine.